### ISSUE TITLE
Protocol specific error code(s): "414", "*", "*".  SQLSTATE=08001 SQLCODE=-30081

### ISSUE THREAD

**Issue by th1460** [ibmdb/python-ibmdb#915]
- Operating System Name: **MacOS M1 Chip System (host) (but running the package `ibm_db` in the podman)**
- DB2
  - Instance Type: Db2 Standalone
  - Machine Type: power
  - Database Version: 11.5.8
- Python Version: 3.9.18 (container)
- ibm_db version: `ibm-db==3.2.1` (container)
- For non-Windows, output of below commands:
  uname (container: Linux, host: Darwin)
  uname -m (contaner: x86_64, host: arm64)
- Test script to reproduce the problem.
- `ibmdb2_connect.py`
```python
import os
import ibm_db

database = os.getenv("DB2_DB")
hostname = os.getenv("DB2_HOST")
username = os.getenv("DB2_USER")
password = os.getenv("DB2_PASSWORD")

conn = ibm_db.connect(
    f"DATABASE={database};"
    + f"HOSTNAME={hostname};"
    + "PORT=50001;"
    + "PROTOCOL=TCPIP;"
    + "SECURITY=SSL;"
    + f"UID={username};"
    + f"PWD={password};"
    + "sslConnection=true;"
    + "sslTrustStoreLocation=trustore.jks",
    "",
    "",
)

connState = ibm_db.active(conn)
print(f"connState: {connState}")
```
- `Dockerfile`
```
FROM python:3.9-bullseye
RUN apt update && apt install -y neovim tmux
COPY requirements.txt .
COPY trustore.jks .
COPY load_archive_ticket.py .
COPY .bashrc /root
COPY .tmux.conf /root
RUN pip install virtualenv
RUN virtualenv venv 
RUN /bin/bash -c "source venv/bin/activate && pip install -r requirements.txt"
```
Steps to Reproduce:
- `podman build --platform linux/amd64 -t ibmdb2 .`
- `podman run --platform linux/amd64 --env-file .env -it ibmdb2 /bin/bash`
- In the container `source venv/bin/activate`
- In the container `python ibmdb2_connect.py`
- Output
```
Traceback (most recent call last):
  File "//load_archive_ticket.py", line 9, in <module>
    conn = ibm_db.connect(
Exception: [IBM][CLI Driver] SQL30081N  A communication error has been detected. Communication protocol being used: "SSL".  Communication API being used: "SOCKETS".  Location where the error was detected: "".  Communication function detecting the error: "sqlccSSLSocketSetup".  Protocol specific error code(s): "414", "*", "*".  SQLSTATE=08001 SQLCODE=-30081
```

PS: with `R` language and `DBeaver` I was able to connect in the same DB:

```r
drv <- RJDBC::JDBC(
  "com.ibm.db2.jcc.DB2Driver",
  "jars/db2jcc4.jar"
)

host <- Sys.getenv("DB2_HOST")
db <- Sys.getenv("DB2_DB")
user <- Sys.getenv("DB2_USER")
password <- Sys.getenv("DB2_PASSWORD")

uri <- sprintf(
  paste0(
    "jdbc:db2://%s:50001/%s:user=%s;password=%s;",
    "sslConnection=true;sslTrustStoreLocation=trustore.jks"
  ),
  host,
  db,
  user,
  password
)

db2 <- DBI::dbConnect(drv, uri)
```

**Comment by imavo** [comment]
Your question states sslTrustStoreLocation=trustore.jks  ( the .jks extension suggests that the file contains a java key store, as used by java/jdbc applications).

python ibm_db is not a java application and  it cannot use a java-keystore (.jks file).

Your options include:  use a .kdb file created with gsk8capicmd_64, or use a PKCS12 format file, or use no keystore.

If you only have a single certificate file , then you do not need a keystore, and instead you can specify the absolute (fully qualified) pathname to the certificate file in your connection-string via the SSLServerCertificate=/path/to/your_certificate_file  method. This is documented in the Db2-LUW knowledge centre.

**Comment by bimalkjha** [comment]
@th1460 Below two keywords used in your connection string is JDBC connection keywords which are not supported for `ibm_db`. ibm_db uses ODBC connection keywords.
```
    + "sslConnection=true;"
    + "sslTrustStoreLocation=trustore.jks",
```
Please remove it. You have already used `"SECURITY=SSL;"` which is equivalent keyword for JDBC keyword `sslConnection=true;`.
You need to generated `*.arm` certificate using GSKit and use it as `SSLServerCertificate=<full path of *.arm file>`. The corresponding ODBC keyword for `sslTrustStoreLocation` is `SSLKeystoreDB` which should point to *.kdb file created by GSKit.
Please check this doc link for three formats of SSL connection string supported by ibm_db: https://github.com/ibmdb/python-ibmdb?tab=readme-ov-file#example-of-ssl-connection-string

Let us know if it resolves your problem. Thanks.

**Comment by th1460** [comment]
I am using the IBM DataMaze how can I generate the `SSLClientKeyStoreDBPassword`?  Because this platform has already provided the `*.kdb` file.

**Comment by bimalkjha** [comment]
@th1460 Get the *.sth file too from the same source who shared *.kdb file and use `SSLClientKeystash=<full path of stash file>;` in connection string. SSLClientKeystoreDBPassword and SSLClientKeystash is mutually exclusive. The keystoredbpassword is the password that we use while creating the *.kdb file. If they can't share password of kdb file with you, get the stash file. Or, ask them to generate SSLServerCerfitificate for client in *.arm form and use it which avoids use of *.kdb and *.sth file. Thanks.

**Comment by th1460** [comment]
@bimalkjha I got the `*.sth` file and I changed the connection:

```
import os
import ibm_db

database = os.getenv("DB2_DB")
hostname = os.getenv("DB2_HOST")
username = os.getenv("DB2_USER")
password = os.getenv("DB2_PASSWORD")

conn = ibm_db.connect(
    f"DATABASE={database};"
    + f"HOSTNAME={hostname};"
    + "PORT=50001;"
    + "PROTOCOL=TCPIP;"
    + "SECURITY=SSL;"
    + f"UID={username};"
    + f"PWD={password};"
    + "SSLClientKeystash=trustore.sth",
    "",
    "",
)

connState = ibm_db.active(conn)
print(f"connState: {connState}")
```
But I get this error:

```
Traceback (most recent call last):
  File "//load_archive_ticket.py", line 9, in <module>
    conn = ibm_db.connect(
Exception: [IBM][CLI Driver] SQL30081N  A communication error has been detected. Communication protocol being used: "SSL".  Communication API being used: "SOCKETS".  Location where the error was detected: "".  Communication function detecting the error: "sqlccSSLSocketSetup".  Protocol specific error code(s): "414", "*", "*".  SQLSTATE=08001 SQLCODE=-30081
```

Now I think this problem is related to *"Incorrectly formatted certificate received from partner"*, correct?

Reference: https://stackoverflow.com/questions/53609339/error-when-connecting-to-db2-with-ibm-db-with-ssl

**Comment by bimalkjha** [comment]
@th1460 You need to use `SSLClientKeyStoreDBPassword` keyword too pointing to full path of *.kdb file along with `SSLClientKeystash` keyword in connection string. If you do not provide *.kdb file in connection string, how driver will know which kdb it has to look for certificate file using the *.sth file? Note that *.sth file stores the encrypted password for the *.kdb file. Driver need password to read from the kdb file.
Basically, your connection string should be as below: 
```
connstr = "DATABASE=BLUDB;HOSTNAME=db2host.ibm.com;PORT=50001;PROTOCOL=TCPIP;SECURITY=SSL;SSLClientKeyStoreDB=/home/user/certs/keystore.kdb;SSLClientKeystash=/home/user/certs/trustore.sth;UID=dbuser;PWD=dbpass"
conn = ibm_db.connect(connstr, "", "")
```
Thanks.

**Comment by th1460** [comment]
Thank you again @bimalkjha a lot for the help. To confirm now, are the keywords correct?

```
import os
import ibm_db

database = os.getenv("DB2_DB")
hostname = os.getenv("DB2_HOST")
username = os.getenv("DB2_USER")
password = os.getenv("DB2_PASSWORD")

conn = ibm_db.connect(
    f"DATABASE={database};"
    + f"HOSTNAME={hostname};"
    + "PORT=50001;"
    + "PROTOCOL=TCPIP;"
    + "SECURITY=SSL;"
    + f"UID={username};"
    + f"PWD={password};"
    + "SSLClientKeystash=trustore.sth"
    + "SSLClientKeyStoreDB=trustore.kdb",
    "",
    "",
)

connState = ibm_db.active(conn)
print(f"connState: {connState}")

```

**Comment by bimalkjha** [comment]
@th1460 The `truststore.sth` and `trustore.kdb` files are for JDBC application and it will not work with non-Java application. You need to download `Secure Connection Certificates.zip` file that come for DB2 CLI and unzip it. After unzipping, you'll get the correct *.kdb and *.sth files for non-Java application as in attached screenshot. You should use these kdb and sth files. Thanks.
![image (1)](https://github.com/ibmdb/python-ibmdb/assets/11818648/43b99ec6-d3e5-4ae0-b8c7-d3c150bd5ab1)

**Comment by bimalkjha** [comment]
Also, you need to use full path of *.sth and *.kdb files as in this connection string. Just using file name will not work:
`database=sample;hostname=hotelfvt04.torolab.ibm.com;port=21170;protocol=TCPIP;Security=SSL;uid=dbuser;pwd=merxxxo;SslClientKeystoredb=/home/bimaljha/ssl/client.kdb;SslClientKeystash=/home/bimaljha/ssl/client.sth` Thanks.

**Comment by th1460** [comment]
The files `trustore.*` and `*.py` that I am running are in `/`:
<img width="638" alt="image" src="https://github.com/ibmdb/python-ibmdb/assets/7317985/22ab69d6-ed8d-4754-a593-e8c78db88346">

**Comment by bimalkjha** [comment]
Looks at the same page for some link mentioning certificate file for ODBC or DB2 CLI applications.. the trustore* files are for Java application. If you are on slack then ping me on bimal-jha1@ibm.com or ask someone in your team to locate the correct certificate files for secure connection. Thanks.

**Comment by bimalkjha** [comment]
The error "414" means GSK_ERROR_BAD_CERT. It clearly says, your certificate file has problem. In your first post, I see this command : `COPY trustore.jks .` How you got trustore.jks file? From where it came in your system?

**Comment by th1460** [comment]
I got by DataMaze. It is a data platform from IBM. There I can download a zip with `*.jks`, `*.kbd` and `*.sth` files.

**Comment by bimalkjha** [comment]
@th1460 Contact DataMaze team and ask for non-java SSL Certificate files generated using GSKit. I don't have idea about DataMaze but if you look on additional pages to find certificates for non-java application, you may get it.
If you have access to IBM Slack, ping me on bimal.jha1@ibm.com so that you can share your screen and we can look for right certificate file. Thanks.

**Comment by th1460** [comment]
Thank you so much for the help. I will see here and I will contact you!
